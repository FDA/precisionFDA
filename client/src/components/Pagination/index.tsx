import React, { useEffect, useState } from 'react'
import styled from 'styled-components'
import { ButtonSolidBlue } from '../Button'
import { inputFocus, InputSelect } from '../form/styles'

export const StyledInputJumpTo = styled.input`
  display: inline-block;
  width: 60px;
  margin-right: 5px;
  padding-left: 10px;
  border: 1px solid #d9d9d9;
  border-radius: 2px;
  transition: all 0.3s;
  ${inputFocus}
`

export const JumpToForm = styled.form`
  display: flex;
`

export const StyledPageOf = styled.div`
  display: inline-block;
  margin-right: 5px;
  white-space: pre;
`

export const StyledPagination = styled.div`
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  gap: 5px;
  font-size: 12px;

  input, select, button {
    font-size: 12px;
  }
`

export const ItemsOf = styled.div`
  margin-right: 5px;
  white-space: pre;
`

export const PerPage = styled.div`
  display: flex;
  align-items: center;
  margin-right: 5px;
  white-space: pre;
`
export const StyledPerPageSelect = styled(InputSelect)`
  margin-right: 5px;
  height: 34px;
`

export function hidePagination(
  isFetched?: boolean,
  length?: number,
  totalPages?: number,
) {
  return length === 0 || (isFetched === false && totalPages === 0)
}

export const Pagination = ({
  page,
  setPage,
  totalCount = 0,
  perPage,
  totalPages = Math.ceil(totalCount / perPage),
  isPreviousData,
  isNextData,
  isHidden,
  onPerPageSelect,
  showPerPage = true,
  showListCount = false,
  showPageOf = true,
  showPageJump = true,
}: {
  page?: number
  setPage: (n: number) => void
  onPerPageSelect: (n: number) => void
  totalCount?: number
  totalPages?: number
  perPage?: number
  isPreviousData: boolean
  isNextData: boolean
  isHidden: boolean
  showListCount?: boolean
  showPerPage?: boolean
  showPageOf?: boolean
  showPageJump?: boolean
}) => {
  if (isHidden) {
    return null
  }
  const [localTotal, setLocalTotal] = useState<number>(totalPages || 0)
  const [localPage, setLocalPage] = useState<number>(page || 0)
  const pageLowerBound = (page - 1) * perPage + 1
  const pageUpperBound = page * perPage

  useEffect(() => {
    if (totalPages) {
      setLocalTotal(totalPages)
    }
  }, [totalPages])

  useEffect(() => {
    if (page) {
      setLocalPage(page)
    }
  }, [page])

  const handleJumpToSubmit = (e: any) => {
    e.preventDefault()
    const value = parseInt(e.target[0].value, 10)
    if (value > localTotal || value <= 0) {
      setPage(1)
    } else {
      setPage(value)
    }
  }

  const handleSetPage = (pa: number) => {
    setLocalPage(pa)
    setPage(pa)
  }

  return (
    <StyledPagination>
      {showListCount && (
        <ItemsOf>{`${pageLowerBound}-${Math.min(
          totalCount,
          pageUpperBound,
        )} of ${totalCount}`}</ItemsOf>
      )}
      {showPerPage && perPage && (
        <PerPage>
          <StyledPerPageSelect
            value={perPage}
            name="perPage"
            onChange={e => onPerPageSelect(parseInt(e.target.value, 10))}
            data-testid="pagination-perpage-select"
          >
            <option value="10">10</option>
            <option value="20">20</option>
            <option value="50">50</option>
            <option value="100">100</option>
          </StyledPerPageSelect>
          Per Page of {totalCount} Total Items
        </PerPage>
      )}
      <ButtonSolidBlue
        data-testid="pagination-previous-page-button"
        type="button"
        onClick={() => handleSetPage(Math.max(localPage - 1, 1))}
        disabled={localPage === 1}
      >
        Previous Page
      </ButtonSolidBlue>{' '}
      <ButtonSolidBlue
        data-testid="pagination-next-page-button"
        type="button"
        onClick={() => handleSetPage(localPage + 1)}
        disabled={localPage === localTotal}
      >
        Next Page
      </ButtonSolidBlue>
      {showPageOf && (
        <StyledPageOf data-testid="pagination-page-of">
          Page {localPage || '...'} of {localTotal || '...'}
        </StyledPageOf>
      )}
      {showPageJump && (
        <JumpToForm onSubmit={handleJumpToSubmit}>
          <StyledInputJumpTo
            data-testid="pagination-jumpto-input"
            defaultValue={1}
            type="number"
            id="page-jumpto"
            name="jumpto"
            max={localTotal}
            min={1}
          />
          <ButtonSolidBlue
            data-testid="pagination-submit-jumpto-button"
            type="submit"
          >
            Jump
          </ButtonSolidBlue>
        </JumpToForm>
      )}
    </StyledPagination>
  )
}
