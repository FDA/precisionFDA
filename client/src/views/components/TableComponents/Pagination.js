import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames/bind'

import { PaginationShape } from '../../shapes/PaginationShape'


const getPages = ({ currentPage, nextPage, prevPage, totalPages }) => {
  const pages = []
  let pagesToShow = 5
  let startFromNumber = 1

  if (prevPage) {
    pages.push({
      label: 'prev',
      value: prevPage,
    })
  }

  if (totalPages <= pagesToShow) {
    pagesToShow = totalPages
  } else {
    if (currentPage + Math.floor((pagesToShow - 1) / 2) >= totalPages) {
      startFromNumber = totalPages - (pagesToShow - 1)
    } else {
      startFromNumber = currentPage - Math.floor(pagesToShow / 2)
    }
  }

  for (let i = 1; i <= pagesToShow; i++) {
    const num = startFromNumber++
    pages.push({
      label: num,
      value: num,
      isActive: num === currentPage,
    })
  }

  if (nextPage) {
    pages.push({
      label: 'next',
      value: nextPage,
    })
  }

  return pages
}

const Page = ({ page, setPageHandler }) => {
  const classes = classNames({
    'pfda-pagination__page--active': page.isActive,
  }, 'pfda-pagination__page')

  const onClick = () => {
    if (page.isActive) return false
    setPageHandler(page.value)
  }

  return (
    <a onClick={onClick} className={classes}>{page.label}</a>
  )
}

const Pagination = ({ data, setPageHandler }) => {
  const showPages = data.totalPages > 1
  return (
    <div className="pfda-pagination">
      {(showPages) && getPages(data).map((page, index) => (
        <Page key={index} page={page} setPageHandler={setPageHandler} />
      ))}
    </div>
  )
}

export default Pagination

Pagination.propTypes = {
  data: PropTypes.exact(PaginationShape),
  setPageHandler: PropTypes.func,
}

Page.propTypes = {
  page: PropTypes.exact({
    label: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    value: PropTypes.number,
    isActive: PropTypes.bool,
  }),
  setPageHandler: PropTypes.func,
}
