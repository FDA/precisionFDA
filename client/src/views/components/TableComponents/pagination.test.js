import React from 'react'
import { shallow } from 'enzyme'

import Pagination from './Pagination'


describe('SpacesListPage test', () => {
  it('should render', () => {
    const data = {
      currentPage: 10,
      nextPage: 9,
      prevPage: 11,
      totalPages: 16,
    }
    const component = shallow(< Pagination data={data} />)
    expect(component).toMatchSnapshot()
  })
})
