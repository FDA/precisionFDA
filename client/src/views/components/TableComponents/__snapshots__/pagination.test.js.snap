// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpacesListPage test should render 1`] = `
<div
  className="pfda-pagination"
>
  <Page
    key="0"
    page={
      Object {
        "label": "<<",
        "value": 11,
      }
    }
  />
  <Page
    key="1"
    page={
      Object {
        "isActive": false,
        "label": 8,
        "value": 8,
      }
    }
  />
  <Page
    key="2"
    page={
      Object {
        "isActive": false,
        "label": 9,
        "value": 9,
      }
    }
  />
  <Page
    key="3"
    page={
      Object {
        "isActive": true,
        "label": 10,
        "value": 10,
      }
    }
  />
  <Page
    key="4"
    page={
      Object {
        "isActive": false,
        "label": 11,
        "value": 11,
      }
    }
  />
  <Page
    key="5"
    page={
      Object {
        "isActive": false,
        "label": 12,
        "value": 12,
      }
    }
  />
  <Page
    key="6"
    page={
      Object {
        "label": ">>",
        "value": 9,
      }
    }
  />
</div>
`;
